### PR TITLE
[bugfix] Metric not logged properly in manual optimization

### LINKED ISSUE #7206: Metric are not logged for Manual Optimization
## 🐛 Bug

When Manual optimaztion is enabled (`automatic_optimization = False`), `self.log("loss", loss)` does not get logged.

Or at least does not get logged to the progress bar for: `self.log("loss", loss, prog_bar=True)`. In my use case it also doesn't get logged to wandb, but I didn't replicate that issue here.

### To Reproduce

Use following [**BoringModel**](https://colab.research.google.com/drive/1nzhu4WP0RBY0-55mAPn_vJBq0iWV16K1?usp=sharing)

1) Run the model, notice that loss doesn't appear in the progress bar
2) Change set `self.automatic_optimization = True` in `BoringModel.__init__`
3) Rerun the model, this type the loss does appear, as expected

### Expected behavior

Metrics logged with `self.log` get logged when `self.automatic_optimization = False`

### Environment
```
* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.8.1+cu101
	- pytorch-lightning: 1.2.10
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.10
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### Additional context

The bug also prevent logging to wandb, but not replicated here

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #7206 #6833

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by kaushikb11 — APPROVED** [review]
LGTM!

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1379,9 +1379,6 @@ def optimizer_step(self, epoch, batch_idx, optimizer, optimizer_idx,
                 optimizer.step(closure=optimizer_closure)
 
         """
-        if not isinstance(optimizer, LightningOptimizer):
```
Comment (tchaton) [inline pytorch_lightning/core/lightning.py:1382]: This is legacy and should have been removed some time ago.

**Review by SeanNaren — APPROVED** [review]

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

File: tests/trainer/optimization/test_manual_optimization.py [hunk 2]
```diff
@@ -1208,3 +1210,74 @@ def training_step(self, batch, batch_idx):
...
+            return optimizer, optimizer_2
+
+    model = TestModel()
+    model.training_epoch_end = None
+    model.val_dataloader = None
+
+    limit_train_batches = 2
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_train_batches=limit_train_batches,
```
Comment (awaelchli) [inline tests/trainer/optimization/test_manual_optimization.py:1267]:
```suggestion
    trainer = Trainer(
        default_root_dir=tmpdir,
        limit_train_batches=2,
```
I don't see it used below

**Review by awaelchli — APPROVED** [review]
LGTM, test fails on master (for 16 bit)

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -1145,9 +1145,9 @@ def call_teardown_hook(self, model: LightningModule) -> None:
         self.teardown(stage=state)
         model.teardown(stage=state)
 
-    def _reset_result_and_set_hook_fx_name(self, hook_name):
+    def _reset_result_and_set_hook_fx_name(self, hook_name: str) -> bool:
         # on_before_zero_grad is called within training_step
-        if "batch_start" in hook_name or "on_before_zero_grad" in hook_name:
+        if "batch_start" in hook_name or hook_name in ["on_before_zero_grad", "on_after_backward"]:
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:1150]:
```suggestion
        if "batch_start" in hook_name or hook_name in ("on_before_zero_grad", "on_after_backward"):
```

**Review by Borda — APPROVED** [review]
